### PR TITLE
Added array typehint support to inspector.

### LINKED ISSUE #19559: Inspector's elements get closed/unfocused when interacted with
**Godot version:** 3.1.dev.custom_build.4da6216

**OS/device including version:** 
Windows 10

**Issue description:** 
ColorPicker closes when you click on everything that is not text box.
ColorPicker from Project Settings works fine.
TextBox gets unfocused when try to type.
TextBox Dialog gets closed but dark overlay remains.

**Steps to reproduce:** 
1. Create 'Label'
2. Add custom dynamic font
3. Expand created font and under 'Settings' click on ColorPicker of 'outline_color'
4. Press on any element of ColorPicker. It closes when button is just pressed.
5. Then try pressing 'Raw Mode'. It closes
6. On second try it works fine.

**Minimal reproduction project:**

[Test.zip](https://github.com/godotengine/godot/files/2102564/Test.zip)

### LINKED ISSUE #20004: Export arrays with type hints do not work (Regression)
**Godot version:**
master @ 8bb90fd5f3b9819bc3bc475e8b9f36c012d6c855

**OS/device including version:**
Windows 10 Home Version 1803 Build 17134.147
Processor i9-8950HK
Nvidia GTX 1050 Ti

**Issue description:**

export(Array, NodePath)  seems to not be working properly in master, the items of the array in the inspector or not being auto-detected as NodePath.
This was working fine in 3.0.2.

**Steps to reproduce:**
1. Create a node with a script with:
```gdscript
export(Array, NodePath) var triggerTargets
```
2. Go to the inspector, and notice that there is no Assign button

3. Clicking the pencil allows you to select NodePath, but this should be automatic because you already defined NodePath in the definition of the export variable

### PR DESCRIPTION
**Pull request by Douglas-W-Williams**
-- supports --
export(Array, ...hint...)
export(Array, ...hint..., ...parameters...) 
export(Array, Array, ...hint...)
export(Array, Array, ...hint..., ...parameters...) 

- [x] FIXED -- export(Array, Color) var myColors - ColorPicker Issue -- Fixes #19559 & #19308
- [x] FIXED -- export(Array, NodePath) var myNodePaths -- Fixes #20004 & #20005
- [x] ADDED -- export(Array, ...hint...)
- [x] ADDED -- export(Array, ...hint..., ...parameters...)
- [x] ADDED -- export(Array, Array, ...hint... )
- [x] ADDED -- export(Array, Array, ...hint..., ...parameters...)

### CODE REVIEW

**Comment by akien-mga** [comment]
@bojidar-bg could maybe help.

**Comment by bojidar-bg** [comment]
`export(Array, ...)` takes the same hints as `export(...)`. So, here are a few examples:
```
export(Array, int, 1, 6) die_rolls # integers between 0 and 6
export(Array, Array, float) heat_map # Array of arrays of floats
export(Array) variants # Array of anything
export(Array, PackedScene) levels # Array of PackedScene resources
enum X {a, b, c}
export(Array, X) xes # array of the X enum, thus with values of 0, 1 or 2 (AFAIR)
export(Array, String, FILE, "*.txt") txtfiles # Array of paths to txt files
```

Basically, everything from the [export docs](http://docs.godotengine.org/en/3.0/getting_started/scripting/gdscript/gdscript_basics.html#exports), but with `Array, ` before the other hints.

**Comment by ordigdug** [comment]
@akien-mga @bojidar-bg All ready for code review.

**Comment by willnationsdev** [comment]
I added your changes to my build and was doing some testing. When I do a direct export of something like this...

    export(Array, String, FILE, GLOBAL) var libs = []

...everything works great! But if I want to add this property to a group by using `_set`, `_get`, and `_get_property_list()` to manually make the PropertyInfo object, it doesn't seem to wanna work. The Inspector shows an `Array[Nil]` value for the property rather than the `Array[0]` value it usually shows (where you can then increase the size to add values and change them). Clicking on it doesn't do anything. I've double-checked to make sure that PropertyInfo fields have the same value as what the typical export statement generates. Here's an example of what I'm doing:

    tool
    extends Resource
    class_name BuildSettings
    var libs = []
    func _get(property):
        match property:
            "settings/libs": return libs
    func _set(property, value):
        match property:
            "settings/libs": libs = property
    func _get_property_list():
        return [
            {
                "name": "settings/libs",
                "type": TYPE_ARRAY,
                "hint": TYPE_STRING_ARRAY,
                "hint_string": str(TYPE_STRING)+"/"+str(PROPERTY_HINT_GLOBAL_FILE)+":*.lib,*.a,*.dll,*.dylib,*.so"
            }
        ]

Is there something I'm doing wrong, or is this a bug that needs fixing?

**Comment by ordigdug** [comment]
@willnationsdev I noticed you deleted your question, did you figure out your issue?

**Comment by willnationsdev** [comment]
@ordigdug yeah, I realized I had the property in my _get_property_list method, but I hadn't realized there was a name mismatch on the _set/_get calls in my code.

File: editor/editor_inspector.h [hunk 1]
```diff
@@ -145,6 +150,7 @@ class EditorProperty : public Container {
 	float get_name_split_ratio() const;
 
 	void set_object_and_property(Object *p_object, const StringName &p_property);
+	void set_parent_object_and_property(Object *p_object, const StringName &p_property);
```
Comment (reduz) [inline editor/editor_inspector.h]: I dont understand what this is used for

File: editor/editor_properties.cpp [hunk 2]
```diff
@@ -1677,19 +1683,28 @@ EditorPropertyColor::EditorPropertyColor() {
...
 	picker->connect("color_changed", this, "_color_changed");
+	picker->connect("popup_closed", this, "_popup_closed");
 }
 
 ////////////// NODE PATH //////////////////////
 
 void EditorPropertyNodePath::_node_selected(const NodePath &p_path) {
 
 	NodePath path = p_path;
-	Node *base_node = Object::cast_to<Node>(get_edited_object());
```
Comment (reduz) [inline editor/editor_properties.cpp]: this seems really hacky, there has to be a better way to do it

File: editor/editor_properties_array_dict.cpp [hunk 3]
```diff
@@ -301,108 +364,387 @@ void EditorPropertyArray::update_property() {
 
 				// atomic types
 				case Variant::BOOL: {
-
 					prop = memnew(EditorPropertyCheck);
 
 				} break;
 				case Variant::INT: {
-					EditorPropertyInteger *ed = memnew(EditorPropertyInteger);
```
Comment (reduz) [inline editor/editor_properties_array_dict.cpp]: This is waay to hacky, too much code that is already in the plugins, if you are going to go full way, just reuse what is there

File: editor/editor_properties.cpp [hunk 4]
```diff
@@ -1714,7 +1729,11 @@ void EditorPropertyNodePath::_node_clear() {
 
 void EditorPropertyNodePath::update_property() {
 
-	NodePath p = get_edited_object()->get(get_edited_property());
```
Comment (reduz) [inline editor/editor_properties.cpp]: Same here, not sure why all this is needed

**Comment by ordigdug** [comment]
From PR meeting reduz stated "as those types of hints will probably be deprecated in favor of a more limited and generic system after 3.1, i would prefer the minimum amount of code is added" 
I'm not going to waste anymore time on this.

**Comment by bojidar-bg** [comment]
@pgruenbacher Bumping old threads is not a good practice, especially when they are mostly unrelated (as is in this case).

**Comment by pgruenbacher** [comment]
yea sorry I was on my phone and meant to do the opposite linking

**Pull request closed without merging**